src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,104 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
         {
             new ReadOnlySpan<T>(array).CopyTo(destination);
         }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second, out elementOffset);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            // note: the calculations below overflow if second.IsEmpty;
+            //       the case second.IsEmpty MUST be handled separately
```

Comment:
@ektrah can you detail why this is the case? At least how it can be the case for both your suggested return values and my proposal that empty spans can return `true`? Not sure I follow how this can be must be true if empty spans can return `true`.

Reply:
Take a look at https://github.com/dotnet/corefx/pull/18731#issuecomment-296856781. `Overlaps` needs to compute `byteOffset + second.Length * sizeof(T) > 2³²` (or 2⁶⁴ on x64) in non-modulo-arithmetic. That can be done in modulo-arithmetic with `byteOffset > -(second.Length * sizeof(T))` for all cases except `second.Length * sizeof(T) == 0`, because 2³²-0 = 2³² and that overflows a 32-bit int.

Do you have an idea for a better way to implement the comparison in modulo-arithmetic?

Reply:
Alternatively, this could be done with `second.Length * sizeof(T) > -(byteOffset)` which works for all cases except `byteOffset == 0`. See https://github.com/dotnet/corefx/pull/24980#issuecomment-341729474 below.